File: game/animation.py
class Animation:
    @staticmethod
    def get_sync_frame(animations, syncs, frame_index):
        if frame_index >= len(syncs):
            frame_index = len(syncs) - 1

        frame = syncs[frame_index] - 1
        if frame < 0:
            print("Animation index less than 0!")
            frame = 0

        if frame >= len(animations):
            print("Animation index out of range!")
            frame = len(animations) - 1

        return animations[frame]

    @staticmethod
    def get_frame(animations, frame_index):
        if frame_index >= len(animations):
            print("Animation index out of range!")
            frame_index = len(animations) - 1

        return animations[frame_index]

File: game-web/src/animation.ts
/** Port of game/animation.py */

export class Animation {
  static getSyncFrame(animations: ImageBitmap[], syncs: number[], frameIndex: number): ImageBitmap {
    if (frameIndex >= syncs.length) frameIndex = syncs.length - 1;
    let frame = syncs[frameIndex] - 1;
    if (frame < 0) { console.warn('Animation index less than 0!'); frame = 0; }
    if (frame >= animations.length) { console.warn('Animation index out of range!'); frame = animations.length - 1; }
    return animations[frame];
  }

  static getFrame(animations: ImageBitmap[], frameIndex: number): ImageBitmap {
    if (frameIndex >= animations.length) {
      console.warn('Animation index out of range!');
      frameIndex = animations.length - 1;
    }
    return animations[frameIndex];
  }
}
